docs/designs/data_transform.md

```diff
@@ -0,0 +1,41 @@
+
+## Transform Expression in SQLFlow
+
+We can extend the SQLFlow syntax and add **TO TRANSFORM** keyword to describe the transform process. Let's take the following SQL expression for example. It trains a model to classify someone's income level using the [census income dataset](https://archive.ics.uci.edu/ml/datasets/Census+Income). The transform expression is **TO TRANSFORM STANDARDIZE(age) as age_std, NORMALIZE(capital_gain) as capital_gain_norm, BUCKETIZED(hours_per_week, bucket_num=10) as hours_per_week_bkt**. It will standardize the column *age* to the column *age_std*, normalize the column *capital_gain* to *capital_gain_norm*, bucketize the column *hours_per_week* to 10 buckets to the column *hours_per_week_bkt*. The output of transform will be passed to the **COLUMN** expression.  
+We add some built-in transform API and users can use them directly in the TRANSFORM expression. The Api set contains NORMALIZE, STANDARDIZE, BUCKETIZED, LOG and more to be added in the future.  
+
+```SQL
+SELECT *
+FROM census_income
+TO TRANSFORM STANDARDIZE(age) as age_std, NORMALIZE(capital_gain) as capital_gain_norm, BUCKETIZED(hours_per_week, bucket_num=10) as hours_per_week_bkt
+TO TRAIN DNNClassifier
+WITH model.hidden_units = [10, 20]
+COLUMNS NUMERIC(age_std), NUMERIC(capital_gain_norm), EMBEDDING(hours_per_week_bkt, dim=128)
+LABEL label
+```
+
+## Design
+
+Data transform contains two stages: analyze and transform. In our design, we will do the analyze using SQL as the first step, and generate the feature column definition as the second step. The feature column contains the transform logic and execute along with the model training process.  
+We choose to convert the **TRANSFORM** expression into two steps of the work flow described by [Couler](https://github.com/sql-machine-learning/sqlflow/blob/develop/python/couler/README.md): analyze and feature column generation. Couler is a programming language for describing workflows. Its compiler translates a workflow represented by a Python program into an [Argo](https://argoproj.github.io/) YAML file. The output of feature column generation will be passed to the next model training step.  
+![data_transform_pipeline](../images/data_transform_pipeline.png)
+
+Let's take STANDARDIZE(age) for example, the following figure describe how the data transform pipeline works in detail.  
+
+![transform_steps](../images/transform_steps.png)
+
+The **Analyze Step** and **Feature Column Generation Step** are two couler steps. Analyze Result and Generated Feature Column Definition Result are the output of these two couler steps.  
```

Comment:
Can we just put the "Analyze Step" in the train step? e.g. for the below SQL statement:

```sql
SELECT * FROM census_income
TO TRAIN DNNClassifier
WITH model.hidden_units = [10, 20]
COLUMNS NUMERIC( STANDARDIZE(age)), ...
LABEL label
```

we can automatically add  analyze field in the actual SELECT statement to MaxCompute or other DB engine: `SELECT *, mean(age) as age_mean, std(age) as age_std FROM ...` then apply `age_mean, age_std` to the transform function.

Reply:
I think SELECT * FROM table is the data retrieving step. It describes which columns we need extract from the source table. Add the aggregate expression in it may be confusing.
What's more, it's may be intuitive to relate the aggregate expression in SELECT with the transform function.